perllib/FixMyStreet/Roles/Translatable.pm

```diff
@@ -0,0 +1,101 @@
+package FixMyStreet::Roles::Translatable;
+
+use Moo::Role;
+
+sub translate_around {
+    my ($orig, $self) = (shift, shift);
+    my $fallback = $self->$orig(@_);
+    (my $col = (caller(2))[3]) =~ s/.*:://;
+    $self->_translate($col, $fallback);
+}
+
+sub translate {
+    my ($self, $col) = (shift, shift);
+    my $fallback = $self->$col(@_);
+    $self->_translate($col, $fallback);
+}
+
+sub _translate {
+    my ($self, $col, $fallback) = @_;
+
+    my $schema = $self->result_source->schema;
+    my $table = lc $self->result_source->source_name;
+    my $id = $self->id;
+
+    # Deal with the fact problem table has denormalized copy of category string
+    if ($table eq 'problem' && $col eq 'category') {
+        my $body_id = $self->bodies_str_ids->[0];
+        return $fallback unless $body_id && $body_id =~ /^[0-9]+$/;
+        my $contact = $schema->resultset("Contact")->find( {
+            body_id => $body_id,
+            category => $fallback,
+        } );
+        return $fallback unless $contact; # Shouldn't happen, but some tests
+        $table = 'contact';
+        $id = $contact->id;
+    }
+
+    if (ref $schema) {
+        my $translation = $schema->resultset('Translation')->find({
+            lang => $schema->lang,
+            tbl => $table,
+            object_id => $id,
+            col => $col
+        });
+        return $translation->msgstr if $translation;
+    } else {
+        warn "Can't use translation on this call to $table.$col";
+    }
+    return $fallback;
+};
+
+sub translation_for {
+    my ($self, $col, $lang) = @_;
+
+    my $schema = $self->result_source->schema;
+
+    my $props = {
+        tbl => lc $self->result_source->source_name,
+        object_id => $self->id,
+        col => $col
+    };
+
+    if ($lang) {
+        $props->{lang} = $lang;
+    }
+
+    my $translations = $schema->resultset('Translation')->search($props);
+
+    return $lang ? $translations->first : $translations;
+}
+
+sub add_translation_for {
+    my ($self, $col, $lang, $msgstr) = @_;
+
+    my $schema = $self->result_source->schema;
+
+    my $props = {
+        tbl => lc $self->result_source->source_name,
+        object_id => $self->id,
+        col => $col,
+        lang => $lang,
+    };
+
+    if ($lang) {
+        $props->{lang} = $lang;
+    }
+
+    my $translation = $schema->resultset('Translation')->find_or_new($props);
```

Comment:
I think here down to the return can be replaced with:
```
$props->{msgstr} = $msgstr; # Actually, put this in the setup above
my $translation = $schema->resultset('Translation')->update_or_create($props, { key => 'translation_tbl_object_id_col_lang_key' });
```
Think that does the same, anyway, worth a check :)